File: workspace/2023/2023-01-27/parameterized-testing.python/parameterized.py
import inspect
import unittest


def parameterized(tuples):
    def new_wrapper(method, tuple):
        return lambda self: method(self, *tuple)

    def decorator(method):
        for i, tuple in enumerate(tuples):
            wrapper = new_wrapper(method, tuple)
            wrapper_name = f"{method.__name__}_{i}"
            f_locals = inspect.currentframe().f_back.f_locals
            f_locals[wrapper_name] = wrapper

    return decorator


class TestCase(unittest.TestCase):

    @parameterized([
        ( 3, 1, 2),
        ( 7, 3, 4),
        (11, 5, 6),
        ( 0, 7, 8),
    ])
    def test(self, expected, a, b):
        self.assertEqual(expected, a + b)


if __name__ == "__main__":
    unittest.main()

File: workspace/2023/2023-06-18/mathematics.ruby/parameterized.rb

def parameterized(name, xs_expected_map)
    xs_expected_map.each_with_index do |(xs, expected), index|
        define_method("test_#{name}_#{index + 1}") do
            actual = method(name)[xs]
            assert_equal expected, actual
        end
    end
end

